File: medium/javascript/2095-delete-middle-node-of-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteMiddle = function (head) {
  let length = 1;
  let cur = head;

  while (cur.next != undefined) {
    cur = cur.next;
    length++;
  }

  const middle = Math.floor(length / 2);

  if (middle === 0) {
    return null;
  }

  cur = head;
  let idx = 0;

  while (idx !== middle - 1 && cur.next != undefined) {
    cur = cur.next;
    idx++;
  }

  const removed = cur.next;
  cur.next = removed.next;
  removed.next = undefined;

  return head;
};

File: medium/go/2095-delete-middle-node-of-linked-list.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteMiddle(head *ListNode) *ListNode {
    length := 1
	cur := head

	for cur.Next != nil {
		cur = cur.Next
		length++
	}

	middle := length/2

	if middle == 0 {
		return nil
	}

	cur = head
	idx := 0

	for idx != middle-1 {
		cur = cur.Next
		idx++
	}

	removed := cur.Next
	cur.Next = removed.Next
	removed.Next = nil

	return head
}